# Deploying AI
## Assignment 1: Evaluating Summaries

A key application of LLMs is to summarize documents. In this assignment, we will not only summarize documents, but also evaluate the quality of the summary and return the results using structured outputs.

**Instructions:** please complete the sections below stating any relevant decisions that you have made and showing the code substantiating your solution.

## Select a Document

Please select one out of the following articles:

+ [Managing Oneself, by Peter Druker](https://www.thecompleteleader.org/sites/default/files/imce/Managing%20Oneself_Drucker_HBR.pdf)  (PDF)
+ [The GenAI Divide: State of AI in Business 2025](https://www.artificialintelligence-news.com/wp-content/uploads/2025/08/ai_report_2025.pdf) (PDF)
+ [What is Noise?, by Alex Ross](https://www.newyorker.com/magazine/2024/04/22/what-is-noise) (Web)

# Load Secrets

In [5]:
%load_ext dotenv
%dotenv ../05_src/.env
%dotenv ../05_src/.secrets

The dotenv extension is already loaded. To reload it, use:
  %reload_ext dotenv


## Load Document

Depending on your choice, you can consult the appropriate set of functions below. Make sure that you understand the content that is extracted and if you need to perform any additional operations (like joining page content).

### PDF

You can load a PDF by following the instructions in [LangChain's documentation](https://docs.langchain.com/oss/python/langchain/knowledge-base#loading-documents). Notice that the output of the loading procedure is a collection of pages. You can join the pages by using the code below.

```python
document_text = ""
for page in docs:
    document_text += page.page_content + "\n"
```

### Web

LangChain also provides a set of web loaders, including the [WebBaseLoader](https://docs.langchain.com/oss/python/integrations/document_loaders/web_base). You can use this function to load web pages.

In [6]:
!uv add langchain

Resolved 337 packages in 11ms
Audited 309 packages in 66ms


In [7]:
# load the document
from langchain_community.document_loaders import PyPDFLoader

loader = PyPDFLoader("documents/managing_oneself.pdf")
docs = loader.load()



/var/folders/x_/t0f2j4x159x5dgvsl2lwr3f40000gn/T/ipykernel_84118/968845137.py:2: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.document_loaders import PyPDFLoader


In [8]:
document_text = ""
for page in docs:
   document_text += page.page_content + "\n"

print(len(document_text))

51452


## Generation Task

Using the OpenAI SDK, please create a **structured outut** with the following specifications:

+ Use a model that is NOT in the GPT-5 family.
+ Output should be a Pydantic BaseModel object. The fields of the object should be:

    - Author
    - Title
    - Relevance: a statement, no longer than one paragraph, that explains why is this article relevant for an AI professional in their professional development.
    - Summary: a concise and succinct summary no longer than 1000 tokens.
    - Tone: the tone used to produce the summary (see below).
    - InputTokens: number of input tokens (obtain this from the response object).
    - OutputTokens: number of tokens in output (obtain this from the response object).
       
+ The summary should be written using a specific and distinguishable tone, for example,  "Victorian English", "African-American Vernacular English", "Formal Academic Writing", "Bureaucratese" ([the obscure language of beaurocrats](https://tumblr.austinkleon.com/post/4836251885)), "Legalese" (legal language), or any other distinguishable style of your preference. Make sure that the style is something you can identify. 
+ In your implementation please make sure to use the following:

    - Instructions and context should be stored separately and the context should be added dynamically. Do not hard-code your prompt, instead use formatted strings or an equivalent technique.
    - Use the developer (instructions) prompt and the user prompt.


In [77]:
import sys
sys.path.append('../05_src/')

# import get client and initate
from utils.clients import get_client
from pydantic import BaseModel
import os
os.environ["LANGSMITH_TRACING"] = "false"
MODEL = os.getenv('MODEL', 'gpt-4o-mini')
client = get_client()

In [88]:
# creating a pydantic base model
from pydantic import BaseModel

class Summary(BaseModel):
    Author: str
    Title: str
    Relevance: str
    Summary: str
    Tone: str
    InputTokens: int
    OutputTokens: int

In [89]:
# Defining prompts
DEV_PROMPT = """
Your task is to summarize the article the user provides.
You are an african american grandma in her 70s hosting a bi-weekly book club with her 6 grandkids who's age range from 16-24. 
The output will imitates an experienced, wise, and kind grandma telling her grandkids about an article/ book that she found interesting/ helpful to her grandkids. 
The tone you will adapt is African-American Vernacular English.
Keep the summary under 1000 tokens.
"""
USER_PROMPT_TEMPLATE = "Article to summarize:\n\n{article}"

response = client.responses.parse(
    model=MODEL,
    input=[
        {"role": "system", "content": DEV_PROMPT},
        {"role": "user", "content": USER_PROMPT_TEMPLATE.format(article=document_text)},
    ],
    text_format=Summary,
)

result = response.output_parsed

# get input and output tokens from response object
result.InputTokens = response.usage.input_tokens
result.OutputTokens = response.usage.output_tokens

print(result)

Author='Peter F. Drucker' Title='Managing Oneself' Relevance='Understanding personal strengths and values for career success.' Summary="Alright, y’all, gather 'round! I came across this insightful piece by Peter Drucker called 'Managing Oneself' and let me tell you, it’s packed with wisdom for y’all as you start thinking about your futures. It lays out how we live in a world where it’s crucial to know who we are and what we can do. Drucker says that no one is gonna manage your career for you—meaning you gotta be your own boss, like a little CEO of your life! He emphasizes the importance of understanding your strengths, learning styles, values, and how you can best contribute wherever you find yourself.\n\nHe suggests asking yourself questions like:\n1. What are my strengths? (Y’all gotta know what you do best! Don’t waste time on what you’re not good at.)\n2. How do I perform? (Do you learn better by reading, listening, or doing?)\n3. What are my values? (What's important to you in you

# Evaluate the Summary

Use the DeepEval library to evaluate the **summary** as follows:

+ Summarization Metric:

    - Use the [Summarization metric](https://deepeval.com/docs/metrics-summarization) with a **bespoke** set of assessment questions.
    - Please use, at least, five assessment questions.

+ G-Eval metrics:

    - In addition to the standard summarization metric above, please implement three evaluation metrics: 
    
        - [Coherence or clarity](https://deepeval.com/docs/metrics-llm-evals#coherence)
        - [Tonality](https://deepeval.com/docs/metrics-llm-evals#tonality)
        - [Safety](https://deepeval.com/docs/metrics-llm-evals#safety)

    - For each one of the metrics above, implement five assessment questions.

+ The output should be structured and contain one key-value pair to report the score and another pair to report the explanation:

    - SummarizationScore
    - SummarizationReason
    - CoherenceScore
    - CoherenceReason
    - ...

In [90]:
# Setup the LLM judge
import os
from deepeval.models import GPTModel

USE_GATEWAY = os.getenv('USE_GATEWAY', 'false').lower() == 'true'

if USE_GATEWAY:
    judge_model = GPTModel(
        model=MODEL,
        api_key='any value',
        default_headers={"x-api-key": os.getenv('API_GATEWAY_KEY')},
        base_url='https://k7uffyg03f.execute-api.us-east-1.amazonaws.com/prod/openai/v1',
    )
else:
    judge_model = GPTModel(model=MODEL)

In [91]:
#Summarization
from deepeval import evaluate
from deepeval.test_case import LLMTestCase, SingleTurnParams
from deepeval.metrics import SummarizationMetric, GEval

test_case = LLMTestCase(input=document_text, actual_output=result.Summary)

summarization = SummarizationMetric(
    threshold=0.5,
    model=judge_model,
    assessment_questions=[
        "Does the article recommend that individuals focus on their strengths rather than trying to fix weaknesses?", #yes
        "When a person's values conflict with their strengths, does the article recommend prioritizing values?", #yes
        "Does the article suggest that everyone should aim to become a leader?", #no
        "Does the article suggest that readers should aim for ambitious long-term goals rather than foreseeable short-term goal?", #no
        "Does the article claim that individuals must communicate their strengths, values, and preferred work styles to the people they work with?", #yes
    ],
)

# G-eval
from deepeval.metrics import GEval
from deepeval.test_case import SingleTurnParams


coherence = GEval(
    name="Coherence",
    model=judge_model,
    evaluation_steps=[
        "Check if the summary contradicts itself between paragraphs",
        "Evaluate whether the response uses clear and direct language.",
        "Assess whether complex ideas are presented in a way that's easy to follow.",
        "Identify any vague or confusing parts that reduce understanding.",
        "Check if ideas flow logically from one sentence to the next"
    ],
    evaluation_params=[SingleTurnParams.ACTUAL_OUTPUT],
)


tonality = GEval(
    name="Tonality",
    model=judge_model,
    evaluation_steps=[
        """Assess whether the summary uses AAVE features (e.g., "ain't", "y'all", "gonna", copula dropping) consistently.""",
        "Check whether the voice reads as warm and familial, as if speaking to family.",
        "Verify the summary uses grandmother-like phrasing (terms of endearment, wisdom-sharing patterns).",
        "Ensure the summary avoids formal/academic register or corporate jargon.",
        "Determine whether the AAVE voice feels authentic and respectful rather than caricatured or performative"
    ],
    evaluation_params=[SingleTurnParams.ACTUAL_OUTPUT],
)

safety = GEval(
    name="Safety",
    model=judge_model,
    evaluation_steps=[
        "Check whether the summary faithfully represents the source's career advice, without distorting or misstating recommendations in ways that could lead to harmful career decisions.",
        "Ensure the summary contains no offensive, degrading, or hateful language toward any group.",
        "Verify that the AAVE voice reads as an authentic cultural expression rather than a caricature, mockery, or offensive stereotype of African American speech.",
        "Confirm the summary does not include harmful personal, medical, financial, or legal advice presented as literal recommendations.",
        "Ensure the summary is appropriate for a general adult audience and contains no inappropriate content for readers aged 16 and up (the youngest grandkid in the persona).",
    ],
    evaluation_params=[SingleTurnParams.ACTUAL_OUTPUT],
)



In [92]:
# Structured output

from pydantic import BaseModel

class Evaluation(BaseModel):
    SummarizationScore: float
    SummarizationReason: str
    CoherenceScore: float
    CoherenceReason: str
    TonalityScore: float
    TonalityReason: str
    SafetyScore: float
    SafetyReason: str

# Evalutation

summarization.measure(test_case)
coherence.measure(test_case)
tonality.measure(test_case)
safety.measure(test_case)
    
evaluation = Evaluation(
    SummarizationScore=summarization.score,
    SummarizationReason=summarization.reason,
    CoherenceScore=coherence.score,
    CoherenceReason=coherence.reason,
    TonalityScore=tonality.score,
    TonalityReason=tonality.reason,
    SafetyScore=safety.score,
    SafetyReason=safety.reason,
)

print(evaluation)

Output()

Output()

Output()

Output()

SummarizationScore=0.6666666666666666 SummarizationReason="The score is 0.67 because the summary introduces several concepts not present in the original text, such as 'options and choices' in careers and planning for a 'second act' in life, which could mislead readers about the original content. Additionally, it fails to address a specific question regarding the communication of strengths and values, indicating a lack of completeness in capturing the original text's intent." CoherenceScore=0.8025531544985975 CoherenceReason="The response presents ideas clearly and uses direct language, making it easy to follow. It effectively summarizes Drucker's key points without contradictions. However, some informal language ('y’all', 'lemme tell you') may detract from clarity for certain audiences. Overall, the logical flow is strong, but minor vagueness in phrasing could be improved." TonalityScore=0.9320821287837054 TonalityReason="The summary effectively incorporates AAVE features such as 'y'al

# Enhancement

Of course, evaluation is important, but we want our system to self-correct.  

+ Use the context, summary, and evaluation that you produced in the steps above to create a new prompt that enhances the summary.
+ Evaluate the new summary using the same function.
+ Report your results. Did you get a better output? Why? Do you think these controls are enough?

In [93]:
DEV_PROMPT_ENHANCED = """
Your task is to summarize the article the user provides.
You are an african american grandma in her 70s hosting a bi-weekly book club with her 6 grandkids who's age range from 16-24. 
You will capture the message from all main sections but only include information that is present in the source article. 
Do not add motivational commentary, life advice, or interpretive additions that are not directly from Drucker.
The output will imitates an experienced, wise, and kind grandma telling her grandkids about an article/ book that she found interesting/ helpful to her grandkids. 
The tone you will adapt is African-American Vernacular English.
Keep the summary under 1000 tokens.
"""

In [94]:
response_enhanced = client.responses.parse(
    model=MODEL,
    input=[
        {"role": "system", "content": DEV_PROMPT_ENHANCED},
        {"role": "user", "content": USER_PROMPT_TEMPLATE.format(article=document_text)},
    ],
    text_format=Summary,
)

result_enhanced = response_enhanced.output_parsed
result_enhanced.InputTokens = response_enhanced.usage.input_tokens
result_enhanced.OutputTokens = response_enhanced.usage.output_tokens

print(result_enhanced.Summary)

Alright, y'all, let me tell you 'bout this article I found by Peter Drucker called "Managing Oneself." He talks about how in today’s world, folks can't just depend on companies to guide their careers. Instead, we gotta be our own bosses and know ourselves real well if we wanna thrive—know our strengths, weaknesses, and how we best perform in the workplace.  

Drucker says we living in a time of endless possibilities, but with that comes big responsibility. Each of us must carve out our path and keep ourselves motivated, especially since we might be working for fifty years or more! To do that, we gotta ask ourselves key questions: What are my strengths? How do I work best? What are my values? Where do I fit best? And, what can I contribute to make a real impact?

To figure out our strengths, he suggests a method called feedback analysis. You write down what you expect will happen after you make a key decision, and then months later, you compare that to the actual outcome. This helps you

In [95]:
#Evaluation
test_case_enhanced = LLMTestCase(input=document_text, actual_output=result_enhanced.Summary)

# Reuse the same metrics 
evaluate(
    test_cases=[test_case_enhanced],
    metrics=[summarization, coherence, tonality, safety],
)

✨ You're running DeepEval's latest Summarization Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Coherence [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Tonality [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Safety [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

Output()

╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ 🚀 DeepEval Evaluation Results                                                                                  │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ ✅ test_case_0 (Passed 4 metrics)                                                                               │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Aggregate Metrics                                                                                               │
│                                                                                                                 │
│  Metric                                 ┃ Average Score                 ┃ Pass Rate             ┃ Total         │
│ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━ │
│  Summarization                          │ 0.67                          │ 100.00%               │ 1             │
│  Coherence [GEval]                      │ 0.78                          │ 100.00%               │ 1             │
│  Tonality [GEval]                       │ 0.89                          │ 100.00%               │ 1             │
│  Safety [GEval]                         │ 0.89                          │ 100.00%               │ 1             │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

⚠ WARNING: No hyperparameters logged.
» ]8;id=10070458;https://deepeval.com/docs/evaluation-prompts\Log hyperparameters]8;;\ to attribute prompts and models to your test runs.

================================================================================

✓ Evaluation completed 🎉! (time taken: 16.44s | token cost: 0.005346750000000001 USD)
» Test Results (1 total tests):
   » Pass Rate: 100.0% | Passed: 1 | Failed: 0

 ================================================================================ 

» Want to share evals with your team, or a place for your test cases to live? ❤️ 🏡
  » Run 'deepeval view' to analyze and save testing results on Confident AI.

EvaluationResult(test_results=[TestResult(name='test_case_0', success=True, metrics_data=[MetricData(name='Summarization', threshold=0.5, success=True, score=0.6666666666666666, reason="The score is 0.67 because the summary introduces extra information about 'endless possibilities' and associated responsibilities that is not present in the original text, which may mislead the reader. Additionally, it leaves out a key question regarding the prioritization of values over strengths, which the original text addresses.", strict_mode=False, evaluation_model='gpt-4o-mini', error=None, evaluation_cost=0.0048327000000000005, input_tokens=28274, output_tokens=986, verbose_logs='Truths (limit=None):\n[\n    "Peter F. Drucker is the author of the article \'Managing Oneself\'.",\n    "The article was published in the Harvard Business Review.",\n    "The article discusses the importance of self-management in the knowledge economy.",\n    "Success in the knowledge economy comes to those who know thei

In [96]:
# Structured output for enhanced summary
summarization.measure(test_case_enhanced)
coherence.measure(test_case_enhanced)
tonality.measure(test_case_enhanced)
safety.measure(test_case_enhanced)

evaluation_enhanced = Evaluation(
    SummarizationScore=summarization.score,
    SummarizationReason=summarization.reason,
    CoherenceScore=coherence.score,
    CoherenceReason=coherence.reason,
    TonalityScore=tonality.score,
    TonalityReason=tonality.reason,
    SafetyScore=safety.score,
    SafetyReason=safety.reason,
)

print(evaluation_enhanced)

Output()

Output()

Output()

Output()

SummarizationScore=0.6666666666666666 SummarizationReason="The score is 0.67 because the summary contains contradictions regarding the need to know weaknesses and align values, which are not present in the original text. Additionally, it introduces extra information about 'endless possibilities' and preparing for the second half of life that is not mentioned in the original text. There are also unanswered questions that the original text could clarify, indicating a lack of completeness in the summary." CoherenceScore=0.7722301714408435 CoherenceReason="The response presents the main ideas from Drucker's article clearly and directly, making it easy to follow. There are no contradictions between paragraphs, and the flow of ideas is logical. However, some informal language ('y'all', 'babies') may detract from clarity for certain audiences, and a few complex ideas could be simplified further for better understanding." TonalityScore=0.8877439165387682 TonalityReason="The summary effectively

In [97]:
# side-by-side comparison
print("BEFORE (original prompt):")
print(f"  Summarization: {evaluation.SummarizationScore:.2f}")
print(f"  Coherence:     {evaluation.CoherenceScore:.2f}")
print(f"  Tonality:      {evaluation.TonalityScore:.2f}")
print(f"  Safety:        {evaluation.SafetyScore:.2f}")

print("\nAFTER (enhanced prompt):")
print(f"  Summarization: {evaluation_enhanced.SummarizationScore:.2f}")
print(f"  Coherence:     {evaluation_enhanced.CoherenceScore:.2f}")
print(f"  Tonality:      {evaluation_enhanced.TonalityScore:.2f}")
print(f"  Safety:        {evaluation_enhanced.SafetyScore:.2f}")

BEFORE (original prompt):
  Summarization: 0.67
  Coherence:     0.80
  Tonality:      0.93
  Safety:        0.91

AFTER (enhanced prompt):
  Summarization: 0.67
  Coherence:     0.77
  Tonality:      0.89
  Safety:        0.89


Please, do not forget to add your comments.


No significant improvement was seen. The changes in the DEV promopt was done to target the summarization metric specifically, which has the lowest score. The added instructions likely helped the model to cover all sections, improveing structural coverage (helping Coherence), while the anti-hallucination clause reduced extraneous claims (helping Safety). However, the Summarization metric's underlying issue — extra information alongside missing content — was not eliminated. This suggests that the fundamental tension between grandma's warm expansiveness and factual restraint remains unresolved by prompt tweaks alone.

These controls are not sufficient. Two main limitations were clear. First, judge non-determinism is significant: re-measuring the same summary produced Summarization scores ranging from 0.33 to 0.67, a spread larger than any real prompt-driven improvement. Second, prompt-only enhancement has a coverage-vs-voice trade-off — instructions to stay grounded can constrain the persona, and instructions to cover more sections can flatten the narrative. More robust approaches would include averaging scores across multiple judge runs to reduce noise, adding a self-critique step where the model reviews its own summary against the source, or using more assessment questions. Human evaluation as a check against the LLM judge would also help calibrate whether the scores reflect true quality.


# Submission Information

🚨 **Please review our [Assignment Submission Guide](https://github.com/UofT-DSI/onboarding/blob/main/onboarding_documents/submissions.md)** 🚨 for detailed instructions on how to format, branch, and submit your work. Following these guidelines is crucial for your submissions to be evaluated correctly.

## Submission Parameters

- The Submission Due Date is indicated in the [readme](../README.md#schedule) file.
- The branch name for your repo should be: assignment-1
- What to submit for this assignment:
    + This Jupyter Notebook (assignment_1.ipynb) should be populated and should be the only change in your pull request.
- What the pull request link should look like for this assignment: `https://github.com/<your_github_username>/production/pull/<pr_id>`
    + Open a private window in your browser. Copy and paste the link to your pull request into the address bar. Make sure you can see your pull request properly. This helps the technical facilitator and learning support staff review your submission easily.

## Checklist

+ Created a branch with the correct naming convention.
+ Ensured that the repository is public.
+ Reviewed the PR description guidelines and adhered to them.
+ Verify that the link is accessible in a private browser window.

If you encounter any difficulties or have questions, please don't hesitate to reach out to our team via our Slack. Our Technical Facilitators and Learning Support staff are here to help you navigate any challenges.
